qiskit/circuit/singleton_gate.py

```diff
@@ -0,0 +1,177 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2023
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""
+Singleton gate classes.
+"""
+import copy
+
+from qiskit.circuit.gate import Gate
+from qiskit.circuit.classicalregister import ClassicalRegister, Clbit
+from qiskit.circuit.exceptions import CircuitError
+
+
+class SingletonGate(Gate):
+    """A base class to use for Gate objects that by default are singleton instances
+
+    This class should be used for gate classes that have fixed definitions and
+    do not contain any unique state. The canonical example of something like
+    this is :class:`~.HGate` which has an immutable definition and any
+    instance of :class:`~.HGate` is the same. Using singleton gates
+    as a base class for these types of gate classes provides a large
+    advantage in the memory footprint of multiple gates.
+
+    The exception to be aware of with this class though are the :class:`~.Gate`
+    attributes :attr:`.label`, :attr:`.condition`, :attr:`.duration`, and
+    :attr:`.unit` which can be set differently for specific instances of gates.
+    For :class:`~.SingletonGate` usage to be sound setting these attributes
+    is not available and they can only be set at creation time. If any of these
+    attributes are used instead of using a single shared global instance of
+    the same gate a new separate instance will be created.
+    """
+
+    _instance = None
+
+    def __new__(cls, *_args, **kwargs):
+        if "label" in kwargs or "_condition" in kwargs or "duration" in kwargs or "unit" in kwargs:
+            return super().__new__(cls)
+        if cls._instance is None:
+            cls._instance = super().__new__(cls)
+        return cls._instance
```

Comment:
How micro-optimising do we want / need to be here?

I'm just thinking that the happy path is nothing in kwargs and the instance already existing, and this code will get called a lot.  I'm wondering if it would have a meaningful effect if we guard the three hashmap lookups in the first `if` behind a `if kwargs and ("label" in kwargs or ...)`, since most of the time, there'll be no kwargs.

I'm also worried a little bit that it's technically possible for something we _expect_ to be a kwarg to be passed positionally such as `XGate("my_label")`, in which case we'll fail to create a mutable instance and then give a weird error.  Certainly the new kwargs could be hidden behind a kwarg-only marker (`class XGate: def __init__(self, label=None, *, duration=...)`).  Solving the pre-existing ones is harder to do efficiently without something gross like overriding `__new__` on all the subclasses or some sort of code-gen in `__init_subclass__`.

Reply:
We should micro-optimize this code to the highest extent possible. This will be called a huge number of times so even a small difference here will impact runtime to a potentially large extent. I'm pretty sure any regressions we saw flagged in the asv numbers above were causedsolely by the extra logic here. Adding a `if kwargs` here is a good idea, I"ll add it.

Reply:
If so, then we may want to go even further and use a little bit of meta-program trickery in `__init_subclass_` (or just manually write it out in all inheritors) that all subclasses of `SingletonGate` need to populate the `_instance` field immediately to let us skip the `if cls._instance is None` check and just always return it.  We should only do that if it actually shows up in benchmarking, though - metaprogramming class inheritance is ugly from a maintenance standpoint (speaking as someone who introduced it to our test suite...).

Reply:
I changed the condition in: https://github.com/Qiskit/qiskit/pull/10314/commits/2479b122cd23264ba1294232b1f9f14d2a08f054

Reply:
I also added handling for args in https://github.com/Qiskit/qiskit/pull/10314/commits/7f99a09c05ff598dd79fa6c0b05e9b5e7f6c81e4 so that if someone passes in a label or other params via positional arguments the returned object is a separate mutable instance.